Multiple --rerun-except arguments are ANDed instead of ORed

While working on #221, I noticed that when passing multiple --rerun-except arguments (or, now when passing a list to `@pytest.mark.flaky(rerun_except=[...])`), all failures are rerun unless they match _all_ of the patterns.

With this test:
``` python
def test_fail():
    raise AssertionError()
```

Running with the example --rerun-except args from the readme gives:
```
$ pytest --reruns 5 --rerun-except AssertionError --rerun-except OSError
======================== test session starts =========================
platform linux -- Python 3.9.2, pytest-7.4.0, pluggy-1.2.0
rootdir: /tmp/test_rerun
plugins: rerunfailures-11.1.2
collected 1 item

test_rerun.py R                                                [100%]R [100%]R [100%]R [100%]R [100%]F [100%]

============================== FAILURES ==============================
_____________________________ test_fail ______________________________

    def test_fail():
>       raise AssertionError()
E       AssertionError

test_rerun.py:2: AssertionError
====================== short test summary info =======================
FAILED test_rerun.py::test_fail - AssertionError
===================== 1 failed, 5 rerun in 0.02s =====================
```

I think the intention in this case (as implied by the docs) is that the test should not be rerun, since the failure matches one of the --rerun-except arguments.

Currently the logic in [_should_hard_fail_on_error](https://github.com/pytest-dev/pytest-rerunfailures/blob/bfc030f841ba5c81493c87ba2a3db3339b500df1/pytest_rerunfailures.py#L318C7-L318C7) is:
- If neither --only-rerun nor --rerun-except are used, all failures are rerun
- Else, if one or more --only-rerun args is given, a failure is rerun if it matches any --only-rerun pattern
- Else, if one or more --rerun-except args is given, a failure rerun if it **fails to match any** --rerun-except pattern

The fix for this isn't exactly obvious, as the intended relationship between --only-rerun and --rerun-except is a bit unclear when they are used together.  My proposal would be:
- If neither --only-rerun nor --rerun-except are used, all failures are rerun (no change)
- Else, if --only-rerun is used alone, a failure is rerun if it matches any --only-rerun pattern (no change)
- Else, if --rerun-except is used alone, a failure is rerun if it matches _none_ of the --rerun-except patterns (new)
- Else, if --only-rerun and --rerun-except are used together, a failure is only rerun if it matches at least one --only-rerun pattern, AND _none_ of the --rerun-except patterns (new)

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -19,8 +19,13 @@ Features
 Bug fixes
 +++++++++
 
--- Fix handling of ``--rerun-except`` and ``--only-rerun`` when used together,
-   and multiple instances of ``--rerun-except``.
+-- Failures are now rerun only if they match at least one ``--only-rerun``
+   pattern (if given) and none of the ``--rerun-except`` patterns. Previously,
+   using both ``--only-rerun`` and ``--rerun-except`` together could cause
+   failures to be rerun even if they did not match any ``--only-rerun``
+   pattern, and when using multiple ``--rerun-except`` patterns, all failures
+   would be rerun unless they matched every pattern.
+   (`#225 <https://github.com/pytest-dev/pytest-rerunfailures/issues/225>`_)
 
 
 11.1.2 (2023-03-09)

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -9,6 +9,11 @@ Features
 
 - Add ``only_rerun`` and ``rerun_except`` arguments to ``@pytest.mark.flaky`` marker.
 
+Bug fixes
++++++++++
+
+-- Fix handling of ``--rerun-except`` and ``--only-rerun`` when used together,
+   and multiple instances of ``--rerun-except``.
 
 11.1.2 (2023-03-09)
 -------------------

--- a/pytest_rerunfailures.py
+++ b/pytest_rerunfailures.py
@@ -315,6 +315,24 @@ def _get_rerun_filter_regex(item, regex_name):
     return regex
 
 
+def _matches_any_rerun_error(rerun_errors, report):
+    for rerun_regex in rerun_errors:
+        try:
+            if re.search(rerun_regex, report.longrepr.reprcrash.message):
+                return True
+        except AttributeError:
+            if re.search(rerun_regex, report.longreprtext):
+                return True
+    return False
+
+
+def _matches_any_rerun_except_error(rerun_except_errors, report):
+    for rerun_regex in rerun_except_errors:
+        if re.search(rerun_regex, report.longrepr.reprcrash.message):
+            return True
+    return False
+
+
 def _should_hard_fail_on_error(item, report):
     if report.outcome != "failed":
         return False
@@ -322,26 +340,25 @@ def _should_hard_fail_on_error(item, report):
     rerun_errors = _get_rerun_filter_regex(item, "only_rerun")
     rerun_except_errors = _get_rerun_filter_regex(item, "rerun_except")
 
-    if not rerun_errors and not rerun_except_errors:
-
+    if (not rerun_errors) and (not rerun_except_errors):
+        # Using neither --only-rerun nor --rerun-except
         return False
 
-    if rerun_errors:
-        for rerun_regex in rerun_errors:
-            try:
-                if re.search(rerun_regex, report.longrepr.reprcrash.message):
-                    return False
-            except AttributeError:
-                if re.search(rerun_regex, report.longreprtext):
-                    return False
+    elif rerun_errors and (not rerun_except_errors):
+        # Using --only-rerun but not --rerun-except
+        return not _matches_any_rerun_error(rerun_errors, report)
 
-    if rerun_except_errors:
-        for rerun_regex in rerun_except_errors:
-            if not re.search(rerun_regex, report.longrepr.reprcrash.message):
+    elif (not rerun_errors) and rerun_except_errors:
+        # Using --rerun-except but not --only-rerun
+        return _matches_any_rerun_except_error(rerun_except_errors, report)
 
-                return False
-
-    return True
+    else:
+        # Using both --only-rerun and --rerun-except
+        matches_rerun_only = _matches_any_rerun_error(rerun_errors, report)
+        matches_rerun_except = _matches_any_rerun_except_error(
+            rerun_except_errors, report
+        )
+        return (not matches_rerun_only) or matches_rerun_except
 
 
 def _should_not_rerun(item, report, reruns):

--- a/test_pytest_rerunfailures.py
+++ b/test_pytest_rerunfailures.py
@@ -575,6 +575,7 @@ def test_no_rerun_on_strict_xfail_with_only_rerun_flag(testdir):
         (["AssertionError: "], True),
         (["ERR"], False),
         (["AssertionError", "OSError"], True),
+        (["ValueError", "OSError"], False),
     ],
 )
 def test_rerun_except_flag(testdir, rerun_except_texts, should_rerun):
@@ -587,7 +588,44 @@ def test_rerun_except_flag(testdir, rerun_except_texts, should_rerun):
 
     pytest_args = ["--reruns", str(num_reruns)]
     for rerun_except_text in rerun_except_texts:
-        print(rerun_except_text)
+        pytest_args.extend(["--rerun-except", rerun_except_text])
+    result = testdir.runpytest(*pytest_args)
+    assert_outcomes(
+        result, passed=num_passed, failed=num_failed, rerun=num_reruns_actual
+    )
+
+
+@pytest.mark.parametrize(
+    "only_rerun_texts, rerun_except_texts, should_rerun",
+    [
+        # Matches --only-rerun, but not --rerun-except (rerun)
+        (["ValueError"], ["Not a Match"], True),
+        (["ValueError", "AssertionError"], ["Not a match", "OSError"], True),
+        # Matches --only-rerun AND --rerun-except (no rerun)
+        (["ValueError"], ["ERR"], False),
+        (["OSError", "ValueError"], ["Not a match", "ERR"], False),
+        # Matches --rerun-except, but not --only-rerun (no rerun)
+        (["OSError", "AssertionError"], ["TypeError", "ValueError"], False),
+        # Matches neither --only-rerun nor --rerun-except (no rerun)
+        (["AssertionError"], ["OSError"], False),
+        # --rerun-except overrides --only-rerun for same arg (no rerun)
+        (["ValueError"], ["ValueError"], False),
+    ],
+)
+def test_rerun_except_and_only_rerun(
+    testdir, rerun_except_texts, only_rerun_texts, should_rerun
+):
+    testdir.makepyfile('def test_only_rerun_except(): raise ValueError("ERR")')
+
+    num_failed = 1
+    num_passed = 0
+    num_reruns = 1
+    num_reruns_actual = num_reruns if should_rerun else 0
+
+    pytest_args = ["--reruns", str(num_reruns)]
+    for only_rerun_text in only_rerun_texts:
+        pytest_args.extend(["--only-rerun", only_rerun_text])
+    for rerun_except_text in rerun_except_texts:
         pytest_args.extend(["--rerun-except", rerun_except_text])
     result = testdir.runpytest(*pytest_args)
     assert_outcomes(
@@ -704,6 +742,7 @@ def test_only_rerun_flag_in_flaky_marker(
         ("ValueError", None, True),
         (["ValueError"], None, True),
         (["OSError", "ValueError"], None, True),
+        (["OSError", "AssertionError"], None, False),
         # CLI override behavior
         ("AssertionError", "ValueError", False),
         ("ValueError", "AssertionError", True),

